src/ore/src/iter.rs

```diff
@@ -41,10 +41,51 @@ where
             Some(v1) => self.all(|v2| v1 == v2),
         }
     }
+
+    /// Converts the the iterator into an `ExactSizeIterator` reporting the given size.
+    ///
+    /// The caller is responsible for providing the correct size of the iterator! Providing an
+    /// incorrect size value will lead to panics and/or incorrect responses to size queries.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the given length is not consistent with this iterator's `size_hint`.
+    fn exact_size(self, len: usize) -> ExactSize<Self> {
+        let (lower, upper) = self.size_hint();
+        assert!(
+            lower <= len && upper.map_or(true, |upper| upper >= len),
+            "provided length {len} inconsistent with `size_hint`: {:?}",
+            (lower, upper)
+        );
+
+        ExactSize { inner: self, len }
+    }
 }
 
 impl<I> IteratorExt for I where I: Iterator {}
 
+/// Iterator type returned by [`IteratorExt::exact_size`].
+#[derive(Debug)]
+pub struct ExactSize<I> {
+    inner: I,
+    len: usize,
+}
+
+impl<I: Iterator> Iterator for ExactSize<I> {
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.len = self.len.saturating_sub(1);
+        self.inner.next()
```

Comment:
Not sure if we want to, but we could enforce the length:

```suggestion
        if self.len > 0 {
            self.len -= 1;
            self.inner.next()
        } else {
            None
        }
```

But, that doesn't protect against not fully consuming the iterator, or an iterator containing more elements than length. We could enforce more in a `Drop` implementation, but that's somewhat redundant with the promis the caller made about the correctness of length.

Reply:
Yeah, silently swallowing updates seems bad too. I'd rather crash loudly if we have a bug. I'd considered something like this:

```
let item = self.inner.next();
if item.is_some() {
    self.len = self.len.checked_sub(1).expect(...);
}
item
```

But I'm not sure if that has performance implications we don't want.